### ISSUE TITLE
Creation of CCEI from scratch

### ISSUE THREAD

**Issue by see24** [LandSciTech/ccviR#25]
Add ability to create CCEI from scratch to prep_data_mod. This would allow different CC scenarios and also different non-breeding ranges, currently restricted to South American nonbreeding ranges.

**Comment by adricaswell** [comment]
Update CCEI exposure table (`tbl_ccei`) to get breaks from the climate read me. Once the climate readme file contains the ccei breaks the code can be updated as follows:
```
    output$tbl_ccei <- gt::render_gt({
      req(spat_res2())
      get_exposure_table(spat_res2(), "CCEI", clim_readme(), clim_readme()$brks_ccei)
    })
```

</br>

Once the value of the CCEI exposure is figured out, this will also need to be added to the table via `get_exposure_table()`. Ideally, the exposure value will be in `spat_res2()` so that the value will be added in the `exp_df` which is created by filtering `spat_res2()` by exp_cave. Then, the `exp_val` associated with `varname == "CCEI"` can be updated to `"ccei_exp_cave"` (or whatever the name of the column containing the CCEI exposure is). Then, the `exp_res_df` should automatically pull the CCEI exposure (like it does for MAT and CMD). If the exposure value will not be in `spat_res2()`, `get_exposure_table()` will require some updating to add the exposure value to the table (it would probably be best to add the exposure value into the `exp_df` as whatever `exp_value` is called).

While doing this, make sure that the breaks are ordered properly. When this table was created we weren't certain if > 7 was class 4 (high) or class 1 (low). Based on comparing the map to the table I think we got it right but worth double checking.

**Comment by see24** [comment]
Lets use https://www.worldclim.org/data/cmip6/cmip6_clim30s.html MAT is in the bioclimatic variables. Will need to figure out how to calculate moisture but I believe it can be done from data available there. I would try to find where https://github.com/bcgov/climr does their creation of climate variables since I know they include CMD

**Comment by steffilazerte** [comment]
For the models to use in an ensemble, looking at the list of 13 climate model ensemble from Mahoney et al 2022 (https://doi.org/10.1002/joc.7566). 

Looking at WorldClim.org data, I realized that in the very small print they refer to the fact that they have variations for some GCMs, so I was able to find more models than we originally thought we had.

set | Model | WorldClim | Notes  | Use
---| -------------------------- | ----------------------------------- |---------- | ------
8 | ACCESS-ESM1.5         | ACCESS-ESM1-5                 |              | ✅ 
13 | BCC-CSM2-MR[^1]   | BCC-CSM2-MR                 | | ✅ 
13 |CanESM5                | CanESM5                   |              | ✅  
8  |CNRM-ESM2-1       | CNRM-ESM2-1                  |              | ✅ 
8  |EC-Earth3               | EC-Earth3-veg                  | with an ["interactive vegetation module"](https://www.cen.uni-hamburg.de/en/icdc/data/cryosphere/cmip6-sea-ice-area/ec-earth3-veg.html) | ❓
8  |GFDL-ESM4            | Not available   for scenarios                 |              | ❌ 
8  |GISS-E2.1-G[^2]               | GISS-E2-1-G (or could use -H)  | | ✅ 
13 | INM-CM5.0            | INM-CM5-0 | | ✅ 
13 | IPSL-CM6A-LR       | IPSL-CM6A-LR | | ✅ 
8  |MIROC6                 | MIROC6 | | ✅ 
8  |MPI-ESM1.2-HR    | MPI-ESM1-2-HR | | ✅ 
8  |MRI-ESM2.0          | MRI-ESM2-0 | | ✅ 
13 | UKESM1-0-LL[^3]                | UKESM1-0-LL | | ✅ 

[^1]: From the Paper, it looks like they used the "-MR" version
[^2]: From the Paper, it looks like they used the "-G" version
[^3]: From the Paper, it looks like they used the "-0-LL" version

So we can use all the models except GFDL-ESM4 which doesn't have the scenarios we need and I'm not sure if EC-Earth3-Veg is still appropriate.

Should we use all the models we can (including EC-Earth3-veg) or just the 11 others, or some other subset?

**Comment by see24** [comment]
Yes I think use the 11 with checkmarks and don't worry about the EC-Earth3-Veg one.

**Comment by steffilazerte** [comment]
**Current status**: 
- Script to produce the CCEI raster: `data-raw/data-ccei.R` (raster output is stored in `misc` which is a non-git tracked folder inside `ccviR`). 
- Functions to summarize the data are in `R/prep_ccei.R`

Still to do
- [ ] Verify the ccei raster is valid
- [ ] Add tests for ccei functions
- [ ] Add ccei raster to Google Drive
- [ ] Test usage of the ccei raster in the App
- [ ] Fix the final CRS from 4326 to 3857 (that used in ccviR in plots, so we don't need to do any transforming in the app)

@see24, I think you could review the main work, understanding that there are still some tweaks to implement. We may need to go through things together, as it gets a bit involved, but hopefully the documentation of the ccei functions will help!

**Comment by see24** [comment]
@steffilazerte I am downloading the ccei files now. (It is running smoothly so far in a background job ❤). Having the tests for the functions would really help with my review. I was thinking it could work to use a small area in `ccei_clip` for testing the whole process.  Maybe New Brunswick? Then we could also save that for the demo files. But it would also be nice to test a small raster with eg 4 quadrants small change in CMD & small change in Tmean, small change in CMD & big change in Tmean, big change in CMD & small change in Tmean, and big change in CMD & big change in Tmean. Where we would expect CCEI to be small, medium, medium, and large.  And maybe do it twice once with low historical variation and once with high? I assume with higher historical variation the CCEI is lower?

**Comment by steffilazerte** [comment]
Make sense, I'll see if I can't get something along those lines figured out soon. As for the variation question... yes, maybe? I'll get back to you on that once I refamiliarize myself with the functions.

**Comment by see24** [comment]
Once the data downloaded I was able to do a quick sanity check against the old NatureServe CCEI raster. It was made for scenario A1B with CMIP3 data so we don't expect them to be exactly the same. 
SSP 2-4.5

![Image](https://github.com/user-attachments/assets/b5b4f338-6b17-4d12-8261-551ec972157b)

 NatureServe A1B

![Image](https://github.com/user-attachments/assets/7baa71a8-e367-43f9-a927-2f1e6808f95a)

I am pretty happy with that! There are some really extreme values in the new CCEI rasters though, including Inf in the 8.5 raster. We should potentially consider capping these or generally look into how they are being created. I wonder if it is areas that previously had no days above 0 so the Eref was 0 before? We just need to change the upper break for ccei in the prep_clim_data function to be the max value in the data, or maybe it can just be Inf?

**Comment by steffilazerte** [comment]
Oh good, I wasn't sure if this was good enough (similar enough). I'll let you know when I get the tests done too (sorry, just putting out fires in some other tests right now 🤦🏼 )

Yes, I agree there were some really high values, I can dig into why that is and then get us a better idea of how to deal with them. I think it was related to variability?

**Comment by see24** [comment]
Happy to report that I was able to successfully get a migratory exposure value with real data using the new CCEI layer. It took me awhile because I forgot that I needed to select a migratory taxonomic group (Eg Bird) as well as check the migratory box and provide both the CCEI layer and the non-breeding range polygon, so just noting those while you are working on tests!. 

One other thing that needs to be done before loading it on to google drive is to:

- [ ] re-run `prep_clim_data` with the new CCEI data

**Comment by steffilazerte** [comment]
I've added tests and tweaked the CCEI functions to be a bit nicer with messaging and easier to test. Part of adding tests was creating a mini set of files (centred on that region in South America with the Inf CCEI values). 

In EXPLORE_CCEI.R I explore a bit these high values. They result from CMD values with SD of 0 (or very low). We could explore either a) Classifying them as you do above, b) setting all SD < 1 as 1 and then classifying as you do above, or c)....? Other highish values seem to result from a historically low CMD (i.e. 0) compared to a future high (e.g., ~400).

I assume that on the Google Drive, we would like the CCEI to be the reclassified version, right? Is that what you were referring to above?

(I'm heading on vacation tomorrow so this will be the last major change for a while. Sorry for the slow going, there were lots of fires to put out over the last two weeks!)

**Comment by see24** [comment]
Yes having the re-classified version on google drive is what I was referring to. I think just changing the classification so that the max for category 4 is Inf would be the simplest and least arbitrary so lets do that. 

I will probably review the changes more thoroughly next week. 

(Enjoy your vacation!!)

**Comment by steffilazerte** [comment]
Okay, I did a couple more things to finalize the CCEI along the lines of what we discussed:

1. I added a bit of code to replace historical SD values of 0 with very low values (otherwise we would get missing values, not Inf, because of dividing by zero). These cells result in very high CCEI's, but that's fine with reclassifying.
2. I created a `data-raw/data_climate_pkg2.R` to start processing the climate data. Right now it just runs `prep_clim_data_multi()` to get the reclassified CCEI's.
3. I uploaded the raw CCEI's (i.e. not reclassified) and the prepared CCEI's to the Google Drive (Supplemental Data > CCEI)

The reclassified CCEI's may need to eventually go in the Climate data package? Or if we have time to setup an alternate data repository (#204) perhaps deal with where it should go then.

**Comment by steffilazerte** [comment]
Next I'll update along the lines of #209

**Comment by steffilazerte** [comment]
Hi @see24 I think this should be  done now! I updated to use climr v0.2.0, and absolute latitudes and re-ran all the CCEI creation scripts. The master files are in Google Drive and the demo files in the package.
